### DISCUSSION TITLE
Ping value calculation in Jamulus

### DISCUSSION THREAD

**Discussion by RepZ97** [jamulussoftware/jamulus#2252] [Documentation]
How does Jamulus calculate the ping value between the client and the server ? Which protocols are being used to gather the network latency information ? Is it icmpv6 ? Thank you.

**Comment by dcorson-ticino-com** [comment]
The client does not calculate the ping time to the servers, it measures the ping time using UDP messages.

**Reply by RepZ97** [reply]
So do the clients exchange ping information using seperate UDP messages or do they calculate based on the same UDP audio stream data ? Like along with the audio payload do they send ping information as well in UDP?

**Reply by hoffie** [reply]
Jamulus Pings are part of the Jamulus UDP protocol: https://github.com/jamulussoftware/jamulus/blob/master/src/protocol.h#L68

They are individual messages (and therefore: packets), but basically they live right next to all other protocol messages as well as the audio data.

Are you attempting to solve some particular issue or asking just out of curiousity?

**Reply by RepZ97** [reply]
Thank you @hoffie & @dcorson-ticino-com for the information. I am actually trying to calculate the end to end latency of a Jamulus audio transmission seperately using wireshark or something similar for a research study that I'm doing. So far no luck with wireshark since UDP packets contain no such latency information which wireshark can observe.

**Reply by dcorson-ticino-com** [reply]
> I am actually trying to calculate the end to end latency of a Jamulus audio transmission seperately

From that description just knowing the digital latency is not enough.
You will have to use audio to do that as the local delay of the audio interface and buffering is not negligible.
I would send an audio blip and measure the time until you hear it come back.  
Pack out your oscilloscope.

**Reply by hoffie** [reply]
> Thank you @hoffie & @dcorson-ticino-com for the information. I am actually trying to calculate the end to end latency of a Jamulus audio transmission seperately using wireshark or something similar for a research study that I'm doing.

Sounds interesting! Would be cool if you keep us posted about any results :)

> So far no luck with wireshark since UDP packets contain no such latency information which wireshark can observe.

Do you know @softins' [wireshark dissector](https://github.com/softins/jamulus-wireshark/) for the Jamulus protocol? I'm pretty sure it'll show the PING messages from the Jamulus protocol.

**Reply by passing** [reply]
> I am actually trying to calculate the end to end latency of a Jamulus audio transmission seperately using wireshark or something similar for a research study that I'm doing. So far no luck with wireshark since UDP packets contain no such latency information which wireshark can observe.

a jamulus ping message has just a 4-byte number as payload:
https://github.com/jamulussoftware/jamulus/blob/master/src/protocol.cpp#L252-L257

the ping is determined in the following way:
- the client sends its own local time value in a ping message
- the server sends the exact same message back to the client
- the client calculates the time difference (time when he got the message back - time that was transmitted in the ping message)

**Reply by RepZ97** [reply]
@dcorson-ticino-com Yes definitely we have to include audio delays as well. I was hoping calculate them seperately after calculating the network delay. Thank you for the idea, and actually.. I might try that to get the absolute end-to-end latency. A very good suggestion. Thank you.

**Reply by RepZ97** [reply]
@hoffie Sure thing. I'd be happy to keep you updated in this thread for future results. Thank you as well for the wireshark dissector suggestion. I'll check that out.

**Reply by RepZ97** [reply]
@passing Thank you for the information. Much appreciated.

**Reply by dcorson-ticino-com** [reply]
> @dcorson-ticino-com I was hoping calculate them seperately after calculating the network delay

Not sure how you can calculate anything in a non-deterministic system.  You may be able to calculate the theoretical minimum, if you know all the characteristics of the parts of the system, but that is not the real world.  Measurement is also subject to the vagarities of the non-deterministic system, but at least it is real and true for the actual case.  If you have the equipment, measurement is also probably much faster to do.

**Reply by RepZ97** [reply]
> > @dcorson-ticino-com I was hoping calculate them seperately after calculating the network delay
> 
> Not sure how you can calculate anything in a non-deterministic system. You may be able to calculate the theoretical minimum, if you know all the characteristics of the parts of the system, but that is not the real world. Measurement is also subject to the vagarities of the non-deterministic system, but at least it is real and true for the actual case. If you have the equipment, measurement is also probably much faster to do.

Well, at least we can calculate the theoretical values for certain certain parameters and conditions (Not all of them. With a margin for indeterminable and negligable delays) and then compare them with our experimental results for those specific parameters and conditions. See the variation of the results against the theoretical values and get a some idea about the standard deviation and what not. Kind of like evaluating the system to see where we can improve.

**Comment by RepZ97** [comment]
Can anyone comment on how the Jamulus client software actually calculate the total audio delay (not the ping) which is displayed on its interface ?

**Reply by pljones** [reply]
Jitter buffers + ping, as far as I'm aware.

**Reply by dcorson-ticino-com** [reply]
See: CClient::EstimatedOverallDelay in client.cpp

**Reply by RepZ97** [reply]
@pljones and @dcorson-ticino-com Thank you. I will check these out.
